### ISSUE TITLE
Opt1 Kernel Development

### ISSUE THREAD

**Issue by ashuping** [johnMamish/nuce468-w22-fluidsim#3]
I'm going to use this issue to keep track of the optimizations we make for the Opt1 Kernel.

**Comment by ashuping** [comment]
Here is the overall idea:

- Instead of four separate kernels, we have just one kernel that performs all steps
- in order to make this work, we have to be smarter about tiling
- Each tile will maintain its own 'old' and 'new' FluidVoxel arrays in shared memory
- A tile will start by loading in each voxel in its range, plus a ring of voxels around the edge
- Tiles will then process these chunks on their own, without needing synchronization

**Comment by ashuping** [comment]
~~Optimization detail: When copying the "old" array into shared memory, if the tile is on an edge/corner, we will need to fill some of the spots in the ring with the "edge condition" voxel to make the stream step work properly. To save computation time, we only actually calculate what that edge condition voxel is if we are on an edge tile.~~

With memcpy_async, we can just do the boundary tile handling while the memory transfer is in progress, saving time without needing to do this.

**Comment by ashuping** [comment]
Optimization detail: Since the collide step does not require any interaction between voxels, we don't actually have to __syncthreads before we move from the array-copy step to the collide step. An individual thread will always complete the array-copy step before it moves onto the collide step, so it will always have the data it needs, even if none of the other warps have copied their voxels yet.

We still need to __syncthreads before we go to the stream step ofc - in that step, voxels need to look at the voxels around them, so all of the warps need to complete the collide step before any of them can stream.

**Comment by ashuping** [comment]
~~Optimization detail: We use cudaMemcpy to handle global memory read/writes. This might help to keep global memory accesses coalesced, since a simple c copy might compile into a series of copies for individual fields, which would end up totally uncoalesced.~~

We ended up implementing our own simple coalesced memcpy.

**Comment by ashuping** [comment]
Optimization detail: To speed up processing, the redundant voxels (the ones in the "halo") will ALSO have threads assigned to calculate their collide step. This means that, for a 32x32 thread-block size, the image will launch thread blocks as if there were only 30x30 threads (i.e. a 60x60 image would take 4 blocks, but a 64x64 image would take 9). This will result in more thread blocks being launched (in some cases), but it will also eliminate the need for threads to handle more than one voxel.

**Comment by ashuping** [comment]
~~Optimization detail: We can use CUDA cooperative groups to make memory access more fine-grained. Basically:~~

1. ~~Block is split into TILE_WIDTHx1-size cooperative groups (we set TILE_WIDTH=32 so that the groups are just warps)~~
2. ~~Each warp retrieves a row of data from global memory (using cudaMemcpyAsync)~~
3. ~~Each warp waits for its row to be done, then starts processing it, without waiting for any other rows to finish~~

We ended up not using CUDA cooperative groups. However, we still use warp-level synchronization to remove the need to synchronize the block between row-copy and collide steps.
